### PR TITLE
Refactor CMake package inclusion

### PR DESCRIPTION
**Pull request by jngrad**
Description of changes:
- move logic to import packages from `CMakeLists.txt` to dedicated helper files `cmake/Find<package>.cmake` for `find_package()`
- enforce the Cython version requested in `CMakeLists.txt`
- CMake now fails if `WITH_CUDA` is set to true but no CUDA-capable compiler is found
- CMake now fails if `WITH_CLANG_TIDY` is set to true but Clang-tidy is not found or its version doesn't match the Clang compiler version
- drop deprecated `FindCUDA` in favor of native CUDA support in CMake 3.10 (required for  #3445)
- add partial support for ROCm 3.1 (closes #3571, required for espressomd/docker#156)

### CODE REVIEW

File: cmake/FindCUDACompiler.cmake [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        get_filename_component(NVCC_EXECUTABLE_DIRNAME "${CUDA_NVCC_EXECUTABLE}" DIRECTORY)
+        get_filename_component(NVCC_EXECUTABLE_DIRNAME "${NVCC_EXECUTABLE_DIRNAME}" DIRECTORY)
+        message(WARNING "Your nvcc (${CUDA_NVCC_EXECUTABLE}) does not appear to match your CUDA libraries (in ${CUDA_TOOLKIT_ROOT_DIR}). While Espresso will still compile, you might get unexpected crashes. Please point CUDA_TOOLKIT_ROOT_DIR to your CUDA toolkit path, e.g. by adding -DCUDA_TOOLKIT_ROOT_DIR=${NVCC_EXECUTABLE_DIRNAME} to your cmake command.")
+      endif()
+      set(CUDA 1)
+      set(CUDA_COMPILER_EXE ${CUDA_NVCC_EXECUTABLE})
+
+      set(CUDA_LINK_LIBRARIES_KEYWORD PUBLIC)
+
+      set(CUDA_NVCC_FLAGS_DEBUG "${CUDA_NVCC_FLAGS_DEBUG} -g")
```
Comment (KaiSzuttor) [inline cmake/FindCUDACompiler.cmake:100]: we should probably introduce a pseudo interface library where we add compile flags and later link against, *e.g.*:
```
add_library(gpu_interface INTERFACE)
target_compile_options(gpu_interface INTERFACE "$<$<CONFIG:DEBUG>:-g>")
```

**Comment by jngrad** [comment]
@KaiSzuttor here is a draft of the next PR where we split the CUDA compiler detection mechanism in separate files: [jngrad/espresso:refactor-cmake-with_gpu](https://github.com/jngrad/espresso/commits/refactor-cmake-with_gpu)

File: maintainer/CI/build_cmake.sh [hunk 2]
```diff
@@ -130,7 +130,10 @@ if [ "${with_coverage}" = true ]; then
 fi
 
 cmake_params="-DCMAKE_BUILD_TYPE=${build_type} -DWARNINGS_ARE_ERRORS=ON -DTEST_NP:INT=${check_procs} ${cmake_params}"
-cmake_params="${cmake_params} -DCMAKE_CXX_FLAGS=${cxx_flags} -DCUDA_NVCC_FLAGS=${nvcc_flags}"
+cmake_params="${cmake_params} -DCMAKE_CXX_FLAGS=${cxx_flags}"
```
Comment (KaiSzuttor) [inline maintainer/CI/build_cmake.sh:133]: why do we manually set cxx flags here?

**Review by KaiSzuttor — COMMENTED** [review]
that improves readability a lot! thanks. Could you please look at my comments, maybe we can get rid of some global flags.

File: cmake/FindCUDACompiler.cmake (hunk 1 above)
Comment (jngrad) [reply cmake/FindCUDACompiler.cmake:100]: we definitely need to do that, but I would prefer to look at it in the next PR, where these flags have been cleaned up and where troubleshooting the CMake logic for different GPU compilers is a lot easier thanks to the new CMake flag

File: maintainer/CI/build_cmake.sh (hunk 2 above)
Comment (jngrad) [reply maintainer/CI/build_cmake.sh:133]: Maybe we can replace that logic with the corresponding `-DCMAKE_BUILD_TYPE=Debug`. This is again overlapping with the next PR, and might also depend on #2977:
https://github.com/espressomd/espresso/blob/be99119c8113b1f2a40ca51d5b07b738bf981eee/.gitlab-ci.yml#L407

Comment (KaiSzuttor) [reply maintainer/CI/build_cmake.sh:133]: all right, let's do it in the other PR

**Review by KaiSzuttor — APPROVED** [review]

**Comment by KaiSzuttor** [comment]
> @KaiSzuttor here is a draft of the next PR where we split the CUDA compiler detection mechanism in separate files: jngrad/espresso:refactor-cmake-with_gpu

just open a draft PR, LGTM

**Pull request merged**